src/formats/pnm.c

```diff
@@ -0,0 +1,222 @@
+// SPDX-License-Identifier: MIT
+// PNM formats decoder
+// Copyright (C) 2023 Abe Wieland <abe.wieland@gmail.com>
+
+#include "loader.h"
+
+#include <setjmp.h>
+
+// Divide a by b, rounding to the nearest integer; evaluates b twice
+#define div_near(a, b) (((a) + (b) / 2) / (b));
+// Divide a by b, rounding up; evalutes b twice
+#define div_ceil(a, b) (((a) + (b) - 1) / (b));
+
+// A file-like abstraction with error handling via setjmp/longjmp which
+// significantly helps code readability
```

Comment:
I agree that in general `setjmp` is terrible for code readability; I only used it here after discovering the alternatives were not great. My main reason for doing so was that we have a fairly small number of errors in parsing, all of which should halt the entire operation, but those errors are detected in different functions. Thus, it actually seemed pretty clean to have errors all land in a single place (which would then print the error and return `ldr_fmterror`) rather than propagating return values.

It also seems reasonable to rewrite `pnm_readc` and `pnm_readuint` to have an error code return and write to pointer arguments, but this would simply make the functions which call them end up having either a lot of repeated code along the lines of
```
size_t v;
int r = pnm_readuint(&it, &v);
if (r) {
    image_print_error(...)
    return ldr_fmterror;
}
```
Honestly I would probably write these checks as local `goto`s anyway (to avoid writing the error condition and return code every time) so a global goto also seemed reasonable, but I can understand your aversion.